Indexed Inequalities on Strict Inequalities (Signed, Unsigned and Float)

SamArch27 opened the pull request:
Indexes rely on weak inequalities (<= and >=) in order to perform range queries.
In Souffle we have strict inequalities (< and >) which need to be translated to weak inequalities to be leveraged in an indexed scan.

Previously, we used the transformation:

Tuple[level, element] < [expr]
into 
Tuple[level, element] <= [expr]-1

This type of transformation poses problems:
1. If <expr> is the lowest or highest value in the domain then an integer overflow can occur when transforming the constraint.
2. For floating-point numbers adding or decrementing to the next smallest/largest value adds complexity.

We simplified things by avoiding indexing strict inequalities for unsigned and float types. However, this places the burden on the user to rewrite all constraints with strict inequalities to weak inequalities. Also, our original technique of adding/subtract 1 to the bound for signed integers is still not safe.

Therefore, in this PR I instead used the following transformation from strict to weak inequalities.

Tuple[level, element] < [expr]
into 
Tuple[level, element] <= [expr] AND Tuple[level, element] != [expr]

In this manner, strict inequalities can still be leveraged to speed up indexed operations and the correct semantics of the constraint are retained without invoking any overflow in arithmetic operations. Users won't have to manually rewrite strict inequality constraints in order to leverage them in an index.

Note:

Currently, evaluation/aggregates4 fails only on interpreter mode only. Otherwise, adopting this new transformation is much cleaner and preferable to what was used previously.

mmcgr commented on tests/interface/functors/libfunctors_la-functors.lo:
```diff
@@ -0,0 +1,12 @@
+# libfunctors_la-functors.lo - a libtool object file
```
Remove this file

b-scholz commented:
It seems that the interpreter is failing on some of the aggregate test-cases. 

This could be a consequence of the interpreter indexes. 

Note that we won't be able to deal with floating point and unsigned inequalities in the interpreter because the lexorder of the interpreter is fixed to a lex-order of signed numbers.

mmcgr commented:
aggregates4 should be added to fastevaluation.at, given it seems to cover a range of aggregates not covered there.

SamArch27 commented:
> It seems that the interpreter is failing on some of the aggregate test-cases.
> 
> This could be a consequence of the interpreter indexes.
> 
> Note that we won't be able to deal with floating point and unsigned inequalities in the interpreter because the lexorder of the interpreter is fixed to a lex-order of signed numbers.

On this particular test case there is an unsigned value of 0xFFFFFFFF and in the query for the sum of all unsigned values < 4 for this relation it ends up including the 0xFFFFFFFF. I assume because when 0xFFFFFFFFF is converted to a signed value it falls in the range.

How can I disable float/unsigned indexed inequalities for the interpreter?

b-scholz commented:
In your inequality transformation you need to check whether the interpreter or the synthesiser will be used via the Global::config(). 

The current idiom is ugly to find out whether we interpret the input program: 
```
 if (!Global::config().has("compile") && !Global::config().has("dl-program") &&
                !Global::config().has("generate") && !Global::config().has("swig")) {
```

SamArch27 commented:
> In your inequality transformation you need to check whether the interpreter or the synthesiser will be used via the Global::config().
> 
> The current idiom is ugly to find out whether we interpret the input program:
> 
> ```
>  if (!Global::config().has("compile") && !Global::config().has("dl-program") &&
>                 !Global::config().has("generate") && !Global::config().has("swig")) {
> ```

I added this check and even after discharging the indexed inequality for non-signed types the test case still fails. The reason this didn't fail in the master build was that the test case that uses unsigned 0xFFFFFFFF was using a strict inequality. However, you can easily make this fail in master on the interpreter by changing this to a weak inequality (<=4). If I had to guess, I think the root cause is that the interpreter evaluates filters and other operations by assuming a RamDomain type for everything. 

Maybe @XiaowenHu96 can take a look at what exactly is going on here.

b-scholz commented:
Can you show me the aggregate and the RAM for which the test fails?

SamArch27 commented:
> Can you show me the aggregate and the RAM for which the test fails?

BEGIN_DEBUG "C(sum  a2 : { A( a2), a2 < 4 }).\nin file /home/sarc9328/souffle/tests/evaluation/aggregates4/aggregates4.dl [25:1-25:35]"
   QUERY
    t0.0=sum t0.0 FOR ALL t0 ∈ A WHERE ((t0.0 != unsigned(4)) AND (t0.0 <= unsigned(4)))
     PROJECT (t0.0) INTO C
  END_DEBUG

^^ Does not work on the interpreter yet:

  BEGIN_DEBUG "C(sum  a2 : { A( a2), a2 < 4 }).\nin file /home/sarc9328/souffle/tests/evaluation/aggregates4/aggregates4.dl [25:1-25:35]"
   QUERY
    t0.0=sum t0.0 FOR ALL t0 ∈ A WHERE (t0.0 < unsigned(4))
     PROJECT (t0.0) INTO C
  END_DEBUG

^^ Works correctly.

@XiaowenHu96 and I worked together to try and figure out what's happening but couldn't figure out anything so far. For the time being, I added a check to the MakeIndexTransformer which doesn't perform the split unless the operation is both indexable and on a signed type. I verified that the type of BinaryConstraintOp was correct and the transformation is from ULT -> ULE + NE. However, Xiaowen found that in the RAM interpreter to the type of the constraint was LE rather than ULE. 

For now, I've added the check and this will pass the test cases but it would be good to figure out the root cause.

b-scholz commented:
Can you write a minimal Datalog Example that shows this behaviour?

b-scholz commented:
We need to fix this. It seems that the semantics of constraint fail for aggregates.

SamArch27 commented:
.decl A(x:unsigned)
A(0xFFFFFFFF).

.decl B(x:unsigned)
.output B

B(x) :- A(x), x <= 10.

This program is failing to produce the correct result (empty) on master and instead, 0xFFFFFFFF gets projected. I don't think it's an aggregate issue since the above program is also failing. However, the issue is only when running the interpreter and not when running the synthesiser.

SamArch27 commented:
> .decl A(x:unsigned)
> A(0xFFFFFFFF).
> 
> .decl B(x:unsigned)
> .output B
> 
> B(x) :- A(x), x <= 10.
> 
> This program is failing to produce the correct result (empty) on master and instead, 0xFFFFFFFF gets projected. I don't think it's an aggregate issue since the above program is also failing. However, the issue is only when running the interpreter and not when running the synthesiser.

RAM:

PROGRAM
 DECLARATION
  A(x:u:unsigned) 
  B(x:u:unsigned) 
 END DECLARATION
 SUBROUTINE stratum_0
  BEGIN_DEBUG "A(0xFFFFFFFF).\nin file /home/sarc9328/benchmarking/testing/test.dl [2:1-2:15]"
   QUERY
    PROJECT (unsigned(4294967295)) INTO A
  END_DEBUG
 END SUBROUTINE
 SUBROUTINE stratum_1
  BEGIN_DEBUG "B(x) :- \n   A(x),\n   x <= 10.\nin file /home/sarc9328/benchmarking/testing/test.dl [7:1-7:23]"
   QUERY
    IF (NOT (A = ∅))
     FOR t0 IN A
      IF (t0.0 <= unsigned(10))
       PROJECT (t0.0) INTO B
  END_DEBUG
  IO B (IO="file",attributeNames="x",name="B",operation="output",output-dir=".",params="{\"records\": {}, \"relation\": {\"arity\": 1, \"auxArity\": 0, \"params\": [\"x\"]}}",types="{\"ADTs\": {}, \"records\": {}, \"relation\": {\"arity\": 1, \"auxArity\": 0, \"types\": [\"u:unsigned\"]}}")
  CLEAR A
 END SUBROUTINE
 BEGIN MAIN
  CALL stratum_0
  CALL stratum_1
 END MAIN
END PROGRAM

XiaowenHu96 commented on src/ram/transform/MakeIndex.cpp:
```diff
@@ -274,6 +122,60 @@ Own<RamCondition> MakeIndexTransformer::constructPattern(const std::vector<std::
+                        indexable = true;
+                        element = lhs->getElement();
+                    }
+                }
```
I don't really know how the transformer works. But I feel like you should include an extra test here:
1.  check if we are in the interpreter mode
2.  check if the target operation is a signed/float one. 
If both true, you should NOT mark `indexable` to true, as it will confuse the aggregates.

mmcgr reviewed (changes requested):
Add aggregates4 to tests/fastevaluation.at. If you're not sure of the format, you can just copy the line for that test from tests/evaluation.at.

(fastevaluation.at is a subset of the evaluation tests that aims to be faster with almost the same coverage. Apparently it does not test much with signed/floats so the initial error was not caught.)

mmcgr commented on tests/fastevaluation.at:
```diff
@@ -6,6 +6,7 @@
 dnl Positive test cases for evaluating Datalog programs
 
+POSITIVE_TEST([aggregates2],[evaluation])
 POSITIVE_TEST([aggregates2],[evaluation])
```
```suggestion
POSITIVE_TEST([aggregates4],[evaluation])
```

XiaowenHu96 commented:
> .decl A(x:unsigned)
> A(0xFFFFFFFF).
> 
> .decl B(x:unsigned)
> .output B
> 
> B(x) :- A(x), x + 1 < 10.
> 
> This program is failing to produce the correct result (empty) on master and instead, 0xFFFFFFFF gets projected. Note that I designed the inequality constraint so that it won't be leveraged by an index scan. I don't think it's an aggregate issue since the above program fails. Note that the issue is only when running the interpreter and not when running the synthesiser.
> 
> Somehow on master when the inequality is indexed it actually works on the interpreter. i.e.
> 
> .decl A(x:unsigned)
> A(0xFFFFFFFF).
> 
> .decl B(x:unsigned)
> .output B
> 
> B(x) :- A(x), x <= 10.
> 
> Will pass on the interpreter on master.
> 
> So I guess if we are doing a filter on the interpreter with unsigned and the value goes outside of the range of RamDomain then it fails.

The behavior is expected. 0xFFFFFFFF u+ 1 overflows to zero. Thats why `x+1 < 1` evalutaed to true but `x < 0` evalulated to false although two ineuqation are actually equivalent.

> Note that the issue is only when running the interpreter and not when running the synthesiser.

 Actually, both interpreter and synthesiser give me the same behavior - maybe do a double check?

SamArch27 commented:
> > .decl A(x:unsigned)
> > A(0xFFFFFFFF).
> > .decl B(x:unsigned)
> > .output B
> > B(x) :- A(x), x + 1 < 10.
> > This program is failing to produce the correct result (empty) on master and instead, 0xFFFFFFFF gets projected. Note that I designed the inequality constraint so that it won't be leveraged by an index scan. I don't think it's an aggregate issue since the above program fails. Note that the issue is only when running the interpreter and not when running the synthesiser.
> > Somehow on master when the inequality is indexed it actually works on the interpreter. i.e.
> > .decl A(x:unsigned)
> > A(0xFFFFFFFF).
> > .decl B(x:unsigned)
> > .output B
> > B(x) :- A(x), x <= 10.
> > Will pass on the interpreter on master.
> > So I guess if we are doing a filter on the interpreter with unsigned and the value goes outside of the range of RamDomain then it fails.
> 
> The behavior is expected. 0xFFFFFFFF u+ 1 overflows to zero. Thats why `x+1 < 1` evaluated to true but `x < 0` evaluated to false although two inequations is actually equivalent. And 0xFFFFFFFF is projected, in the form of unsigned.
> 
> > Note that the issue is only when running the interpreter and not when running the synthesiser.
> 
> Actually, both interpreter and synthesiser give me the same behavior - maybe do a double check?

Yep this is my bad. I must have had the wrong binary like you said.

The issue persists for me without guarding for the interpreter in MakeIndex for this branch even though I've triple checked to make sure that  my transformation is doing the right thing. I've probably done something silly but it's taking me a while to figure out what's gone wrong.

mmcgr commented:
Are you running `make distclean` to be sure none of the built products remain?

SamArch27 commented:
> Are you running `make distclean` to be sure none of the built products remain?

Thanks for the suggestion. It didn't seem to change anything but it's good to know this command now.

To clarify, the PR passes all the tests because I'm specifically doing a check to not split < into <= and != on the interpreter. If I remove this check then aggregates4 will fail.

I'm still really unsure what is causing this failure because as far as I understand the transformation is being performed correctly.

SamArch27 commented:
Update:

I read the email from Xiaowen and I think it's the most sensible idea.

To resolve all issues we can simply approve this PR as is. Effectively this PR:

1. Ensures that no inequalities are indexed for unsigned/float attributes in the interpreter
2. Stops all strict inequalities from being directly indexed 
3. Converts strict inequalities that could be indexed into weak inequalities (e.g. < to <= and !=) either when: 
a) The synthesiser is being used
b) The interpreter is being used AND the type of the attribute is signed

Let me know if there are any further questions.
